UnitTestFrameworkPkg/Library/Posix/MemoryAllocationLibPosix/AllocateBelowAddress.c

```diff
@@ -0,0 +1,435 @@
+VOID *
+HostAllocateBufferInRange (
+  UINTN  MaximumAddress,
+  UINTN  Length
+  )
+{
+  UINTN  Address;
+  VOID   *AllocatedAddress;
+
+  if (mPageSize == 0) {
+    mPageSize = HostGetPageSize ();
+    if (mPageSize == 0) {
+      return NULL;
+    }
+  }
+
+  //
+  // Round maximum address down to the nearest page boundary
+  //
+  MaximumAddress &= ~(mPageSize - 1);
+
+  for (Address = MaximumAddress; Address >= MINIMUM_ALLOCATION_ADDRESS; Address -= mPageSize) {
+ #if defined (_WIN32) || defined (_WIN64)
+    AllocatedAddress = VirtualAlloc (
+                         (VOID *)Address,
+                         Length,
+                         MEM_RESERVE | MEM_COMMIT,
+                         PAGE_READWRITE
+                         );
+    if (AllocatedAddress != NULL) {
+      return AllocatedAddress;
+    }
+
+ #elif defined (__linux__)
+    AllocatedAddress = mmap (
+                         (VOID *)Address,
+                         Length,
+                         PROT_READ | PROT_WRITE,
+                         MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE,
+                         -1,
+                         0
+                         );
+    if (AllocatedAddress != MAP_FAILED) {
+      return AllocatedAddress;
+    }
+
+ #else
+    return NULL;
+ #endif
+  }
+
+  return NULL;
+}
+
+/**
+  Use system services to free memory allocated with HostAllocateBufferInRange().
+
+  @param[in] Buffer  Pointer to buffer previously allocated with
+                     HostAllocateBufferInRange().
+  @param[in] Length  Length, in bytes, of buffer previously allocated with
+                     HostAllocateBufferInRange().
+**/
+static
+VOID
+HostFreeBufferInRange (
+  IN VOID   *Buffer,
+  IN UINTN  Length
+  )
+{
+ #if defined (_WIN32) || defined (_WIN64)
+  if (!VirtualFree (Buffer, 0, MEM_RELEASE)) {
+    ASSERT (FALSE);
+  }
+
+ #elif defined (__linux__)
+  if (munmap (Buffer, Length) == -1) {
+    ASSERT (FALSE);
+  }
+
+ #endif
+}
+
+/**
+  Allocate memory below a specifies address.
+
+  @param[in] MaximumAddress  The address below which the memory allocation must
+                             be performed.
+  @param[in] Length          The size, in bytes, of the memory allocation.
+
+  @retval  !NULL  Pointer to the allocated memory.
+  @retval  NULL   The memory allocation failed.
+**/
+VOID *
+EFIAPI
+HostAllocatePoolBelowAddress (
+  IN UINT64  MaximumAddress,
+  IN UINT64  Length
+  )
+{
+  VOID                     *AllocatedAddress;
+  POOL_HEAD_BELOW_ADDRESS  *PoolHead;
+
+  if (Length == 0) {
+    return NULL;
+  }
+
+  //
+  // Limit maximum address to the largest supported virtual address
+  //
+  MaximumAddress = MIN (MaximumAddress, MAX_UINTN);
+
+  //
+  // Increase requested allocation length by the size of the pool header
+  //
+  Length += sizeof (POOL_HEAD_BELOW_ADDRESS);
+
+  //
+  // Make sure allocation length is smaller than maximum address
+  //
+  if (Length > MaximumAddress) {
+    return NULL;
+  }

```

Comment:
An assert that outputs the details might be valuable when returning `NULL` in this case.

```suggestion
  if (Length > MaximumAddress) {
    ASSERT (Length <= MaximumAddress);
    return NULL;
  }
```

Reply:
Unlike the NULL checks for preconditions, this is ASSERT abuse. ASSERTs are supposed to be  tautologies (considering the reachability condition). In AUDK (Acidanthera’s EDK II fork), we introduced DEBUG_RAISE for such debugging aids, which can be turned on and off separately from ASSERTs, such that fuzz-testing and unit tests for failure do not trigger them. If there still should be an ASSERT, it should at least be ASSERT(FALSE) and documented, a locally-unsat ASSERT without a comment is irritating,

Reply:
I agree that assert might be overboard here. My main point is that a debugging aid might be helpful.

Reply:
How about a DEBUG_ERROR message?

Reply:
This is another thing that might need proper discussion and actually executed solution. Every time a "this is technically an error that is to be gracefully handled, but it's still super surprising and needs dev attention", the points are that an ASSERT is overkill and a debug message hat can be missed is not enough. As I said, in AUDK, we have DEBUG_RAISE as a middle-ground. Also, our OpenCore boot loader allows to optionally halt execution when a DEBUG_ERROR is logged.